src/child.ml

```diff
@@ -204,17 +217,34 @@ let run state comms_sock fd_sock fd_sock_path =
 			List.iter (fun fd -> Unix.close fd) fds;
 
 			(* Close the end of the pipe that's only supposed to be written to by the child process. *)
-			Opt.iter Unix.close !out_childlogging;
-
-			Opt.iter
-				(fun in_fd ->
-					let key = (match state.syslog_stdout.key with None -> Filename.basename name | Some key -> key) in
-					(* Read from the child's stdout and write each one to syslog *)
-					Unixext.lines_iter
-						(fun line ->
-							Fe_debug.info "%s[%d]: %s" key result line
-						) (Unix.in_channel_of_descr in_fd)
-				) !in_childlogging;
+			Opt.iter Unix.close !out_childlogging_stdout;
+			Opt.iter Unix.close !out_childlogging_stderr;
+
+			let running = ref true in
+			let send_log = ref Fe_debug.info in
+			let log_fd = ref state.syslog_stdout.key in
+			let key = (match !log_fd with None -> Filename.basename name | Some key -> key)
+			in
+			while !running do
+				let (in_fd, _, _) =
+					Unix.select [Opt.unbox !in_childlogging_stdout; Opt.unbox !in_childlogging_stderr] [] [] (-1.0)
+				in
+
+				List.map (fun fd ->
+					if fd = Opt.unbox !in_childlogging_stderr then
+					begin
+						log_fd := state.syslog_stderr.key;
+						send_log := Fe_debug.error
+					end;
+
+					let line = try Some (input_line (Unix.in_channel_of_descr fd))
+						with End_of_file -> None
+					in
+					match line with
+					| Some line -> !send_log "%s[%d]: %s" key result line
+					| None -> running := false
+				) in_fd;
+			done;
```

Comment:
The loop terminates, when one the file descriptors is closed. What about the log messages in the other file descriptor? The handling of the lines read can be simplified with much less pattern matching:

    try
      let line = input_line (Unix.in_channel_of_descr fd) in
      !send_log "%s[%d]: %s" key result line
    with
      End_of_file -> running := false

Overall, this design needs to be re-thought.